pygen/pygen_src/riscv_instr_sequence.py

```diff
@@ -1,89 +1,191 @@
+
+        if not is_main_program:
+            self.gen_stack_enter_instr()
+            self.gen_stack_exit_instr()
+
+    # TODO
+    def gen_stack_enter_instr(self):
+        pass
+
+    # TODO
+    def gen_stack_exit_instr(self):
+        pass
+
+    '''
+    ----------------------------------------------------------------------------------------------
+    Instruction post-process
+
+    Post-process is required for branch instructions:
+
+    Need to assign a valid branch target. This is done by picking a random instruction label in
+    this sequence and assigning to the branch instruction. All the non-atomic instructions
+    will have a unique numeric label as the local branch target identifier.
+    The atomic instruction streams don't have labels except for the first instruction. This is
+    to avoid branching into an atomic instruction stream which breaks its atomicy. The
+    definition of an atomic instruction stream here is a sequence of instructions which must be
+    executed in-order.
+    In this sequence, only forward branch is handled. The backward branch target is implemented
+    in a dedicated loop instruction sequence. Randomly choosing a backward branch target could
+    lead to dead loops in the absence of proper loop exiting conditions.
+    ----------------------------------------------------------------------------------------------
+    '''
+
+    def post_process_instr(self):
+        label_idx = 0
+        branch_cnt = 0
+        j = 0
+
+        def def_val():
+            return 0
+        branch_target = defaultdict(def_val)
+
+        for instr in self.directed_instr:
+            self.instr_stream.insert_instr_stream(instr.instr_list)
+        '''
+        Assign an index for all instructions, these indexes wont change
+        even a new instruction is injected in the post process.
+        '''
+        for i in range(len(self.instr_stream.instr_list)):
+            self.instr_stream.instr_list[i].idx = label_idx
+            if(self.instr_stream.instr_list[i].has_label and
+                    not(self.instr_stream.instr_list[i].atomic)):
+                if((self.illegal_instr_pct > 0) and
+                   (self.instr_stream.instr_list[i].insert_illegal_instr == 0)):
+                    '''
+                    The illegal instruction generator always increase PC by 4 when resume execution,
+                    need to make sure PC + 4 is at the correct instruction boundary.
+                    '''
+                    if(self.instr_stream.instr_list[i].is_compressed):
+                        if(i < (len(self.instr_stream.instr_list) - 1)):
+                            if(self.instr_stream.instr_list[i + 1].is_compressed):
+                                self.instr_stream.instr_list[i].is_illegal_instr = random.randrange(
+                                    0, min(100, self.illegal_instr_pct))
+                        else:
```

Comment:
the `else` statement is aligned with the wrong `if` statement, this should be aligned with the `if(self.instr_stream.instr_list[i].is_compressed)`